### DISCUSSION TITLE
2.0.0-rc2 (2023-07-10)

### DISCUSSION THREAD

**Discussion by brendanhay** [brendanhay/amazonka#927] [Announcements]
The Amazonka team would like to announce the availability of the final release candidate for the upcoming `2.0.0` release. This release contains many more bug fixes and AWS service updates since [rc1](#716), including:

- An overhaul of the [authentication-related code](https://github.com/brendanhay/amazonka/issues/746). 
- A new [hooks system](https://github.com/brendanhay/amazonka/issues/875) for clients to inject custom behavior into the request/response lifecycle.
- Nix [Flakes](https://nixos.wiki/wiki/Flakes) (with `cabal`) are used for development of the library.
- The `cryptonite` dependency has been replaced with [crypton](https://github.com/brendanhay/amazonka/issues/920).

Please read the [Amazonka CHANGELOG.md](https://github.com/brendanhay/amazonka/blob/main/lib/amazonka/CHANGELOG.md) for a high-level overview of all the other changes included in the release candidate.

To test the release, you will need to add a [`source-repository-package`](https://cabal.readthedocs.io/en/3.4/cabal-project.html#specifying-packages-from-remote-version-control-locations) section to your `cabal.project` file for `amazonka` itself, and for any service libraries you use (`amazonka-sts` and `amazonka-sso` are now dependencies of `amazonka` because of `sts:AssumeRoleWithWebIdentity` and SSO support):

```
-- For amazonka
-- Multiple subdirs in a single `source-repository-package` stanza are supported by cabal-install >= 3.2.0.0.
source-repository-package
    type: git
    location: https://github.com/brendanhay/amazonka
    tag: 31f3d044d84912b99e396a97085a9678e699d91f
    subdir: lib/amazonka lib/amazonka-core lib/services/amazonka-sso lib/services/amazonka-sts lib/services/amazonka-<servicename>
```

Stack users should use an [`extra-deps:`](https://docs.haskellstack.org/en/stable/pantry/) stanza like this:
```yaml
extra-deps:
- github: brendanhay/amazonka
  commit: 31f3d044d84912b99e396a97085a9678e699d91f
  subdirs:
  - lib/amazonka
  - lib/amazonka-core
  - lib/services/amazonka-sso
  - lib/services/amazonka-sts
  - lib/services/amazonka-<servicename>
```

Stack users who also use `haskell.nix` will need to tell `haskell.nix` which branch to fetch Amazonka commits from by passing a `branchMap` parameter like this:
```nix
pkgs.haskell-nix.project {
  branchMap = {
    "https://github.com/brendanhay/amazonka.git" = {
      31f3d044d84912b99e396a97085a9678e699d91f = "main";
    };
  };

  # Whatever else you normally have here
  src = ...
}
```

Please file any feedback from testing under the related [discussion](https://github.com/brendanhay/amazonka/discussions/927) so the team can finalize the `2.0.0` release.

**Comment by endgame** [comment]
Excellent news, thanks for cutting the release. We will have to wait for a few packages to catch up before I can add `aeson-2.2` support to amazonka and there are a few final chores to do, but this is a big milestone on the road to Hackage.

**Comment by axman6** [comment]
Whooohooo, well done team! I'll have to go update my things!

In thew stack example above, can 
```
commit: <current revision of the `main` branch>
```
be replaced with an actual commit that is _the_ RC2 commit? I assume it should be `31f3d044d84912b99e396a97085a9678e699d91f` from the Nix file?

**Reply by brendanhay** [reply]
Yeah, I'll update the release notes to replace the placeholder so it's not ambiguous.

**Reply by endgame** [reply]
And the `branchMap` entry as well, I guess.

**Reply by brendanhay** [reply]
That one was already correct - they're both the same SHA as the release tag.

**Comment by axman6** [comment]
Also it looks like the links to the `authentication-related code`, `hooks system` etc. haven't worked, because GitHub is janky.

**Reply by brendanhay** [reply]
Seems like absolute `https://` urls are needed rather than `#<issue-number>` shorthand in Discussions, which, is amazing.

**Reply by axman6** [reply]
Sure is! yeah I could immediately see the reasonable thing you'd done and the unreasonable breakage GitHub had injected.

**Reply by brendanhay** [reply]
The urls are updated here and also in the discussion, because doing everything in two separate places is the way.

**Comment by tfausak** [comment]
Would it be possible to add the package tarballs (like `amazonka-s3-2.0.0.tar.gz`) to the release? That way it would be possible to use them without having to clone the entire repository.

**Reply by brendanhay** [reply]
> Successfully uploaded 338 assets to 2.0.0-rc2

Had to hand crank generator all night keep the rural NZ internet alive, should be attached to the release under "Assets" now.

@endgame I've added `scripts/release-tarballs`  as the potential beginnings of a tag/upload release script. TBD if we want to commit to uploading art{i,e}facts as part of all future releases - it likely isn't necessary once Hackage uploads are in a "healthy" place again.

**Reply by endgame** [reply]
I think doing it case-by-case is fine. I don't see many projects using release tarballs because Hackage has that covered. Depending on how things go with the large-scale generator reword i'll have to do post-2.0, we might have to do it for a tag here or there.

**Reply by tfausak** [reply]
I thought this would allow Cabal users to use the package tarballs from GitHub directly, but unfortunately a Cabal bug makes them get downloaded every time: https://github.com/haskell/cabal/issues/9113

However it's still convenient to manually download the tarball and stuff it in a `vendor/` directory or something. Thanks!

**Comment by endgame** [comment]
I finished the 2.0 milestone today, but I'd like to get work running on latest `main` just to be absolutely sure things are working before we hit the big red button, just in case making `amazonka` and `amazonka-core` hlint-clean has caused some subtle breakage.

**Reply by endgame** [reply]
It is probably also worth giving it one more week so people actually have a chance to play with RC2.

**Reply by endgame** [reply]
@brendanhay Can you please look at CI if you get a chance? I see a bunch of GHC panics which I can't reproduce locally, and maybe a cache (which I don't have access to) needs flushing?

**Reply by brendanhay** [reply]
Sure. I've deleted the GH caches and re-triggered the latest `main` merge/workflow, lets' see how it goes.

**Reply by endgame** [reply]
[Seems to have worked](https://github.com/brendanhay/amazonka/actions/runs/5582344962).

CI is good, work is now running off latest `main` (so the hlint fixes, aeson-2.2 bounds, etc), and all seems to be fine. When do we want to finalise the changelog, build the sdists, tag the release, and send the lot to Hackage? 2023-07-24 would be a two-week RC period. I would hope that lots of people have been on fairly recent `main` at this point, so bumping and testing RC2 should not have been a big ask.

**Reply by endgame** [reply]
It's probably also worth having a 2.0 migration guide somewhere that isn't split across the RC1 and RC2 announcements. Any suggestions for where you might want that?

**Reply by brendanhay** [reply]
I'd put it in a release discussion (same as this thread) and also a possibly elided form (or pointer/link) into the `CHANGELOG.md` so it shows up on Hackage.

**Reply by brendanhay** [reply]
Possibly also update the `{cabal,readme}.ede` templates to point back to instructions in this repository just so the migration notes show up for each package when you hit `https://hackage.haskell.org/package/amazonka-*`.

**Reply by endgame** [reply]
I was thinking a section in the main `amazonka` readme with pointers to the relevant changelog sections: that would show up on Hackage and means we're not beholden to GitHub any more than we already are. Does that sound OK to you?

I should be able to do this in the next couple of days.

**Reply by endgame** [reply]
@brendanhay The migration guide has been written and merged, and work is running on latest `main` (including the EC2 metadata client changes). If you're happy, then I think it's time to tie it all off (update changelog date, build sdists, create tag), and ship the sdists to Hackage. Are you able to do that?

**Reply by brendanhay** [reply]
See https://github.com/brendanhay/amazonka/releases/tag/2.0.0

**Comment by tfausak** [comment]
For what it's worth, we upgraded from RC1 to RC2 at work and it went smoothly. Thanks for all your hard work, @endgame & @brendanhay!

**Comment by jhrcek** [comment]
We upgraded our codebase from RC1 to RC2 2 weeks ago. Since then it's been running in production without any issues.
I appreciate your work on this @endgame and @brendanhay :heart:
